internal/jujucommands/jujucommands.go

```diff
@@ -5,127 +5,86 @@
-// Upon receiving an error in an OutputLine, the line previously streamed is guaranteed to be the error line written by
-// the command.
-// I.e., if you run: "help -b"
-// You will get:
-//
-//	Output line: ERROR option provided but not defined: -b // The error line from juju (Which is a none-erroneous OutputLine)
-//	Command finished with error: cmd failed with code 2 // The command error code line (Contains an error in the OutputLine)
-//
-// Consumers are expected to simply read from the OutputLine channel and check for an error like so:
-//
-//	for out := range outputCh {
-//		if out.Err != nil {
-//			fmt.Println("Command finished with error:", out.Err)
-//			break
-//		}
-//		fmt.Println("Output line:", out.Line)
-//	}
-//
-// The OutputLine channel is closed once all output from the command has finished an the error has been captured.
-//
-// Lines returned are returned with no newlines.
-func runCmdWithOutputRetriever(store jujuclient.ClientStore, cmdAndArgs string) (<-chan outputLine, error) {
-	cmdReader, cmdWriter := io.Pipe()
-
-	cmdCtx, err := cmd.DefaultContext()
+// runJujuCmd runs a juju command with the given command string and JUJU_DATA directory.
+// It returns a channel that will receive output lines from the command's stdout and stderr.
+// The command is run in a separate goroutine, and the context can be used to cancel the command.
+func runJujuCmd(ctx context.Context, cmdStr, jujuDataDir string) (<-chan outputLine, error) {
+	env := func(key string) string {
+		if key == "JUJU_DATA" {
+			return jujuDataDir
+		}
+		return ""
+	}
+	args, err := shell.Fields(cmdStr, env)
+	if err != nil {
+		return nil, fmt.Errorf("failed to parse command: %w", err)
+	}
+
+	cmd := exec.Command("juju", args...)
+
+	stdOut, err := cmd.StdoutPipe()
+	if err != nil {
+		return nil, fmt.Errorf("failed to get stdout: %w", err)
+	}
+
+	stdErr, err := cmd.StderrPipe()
 	if err != nil {
-		return nil, err
+		return nil, fmt.Errorf("failed to get stderr: %w", err)
 	}
 
-	cmdCtx.Stderr = cmdWriter
-	cmdCtx.Stdout = cmdWriter
-
-	outputCh := make(chan outputLine)
-	// We buffer cmdFinishedCh (capacity 1) to avoid deadlock.
-	// runCmd defers cmdWriter.Close(), which only runs after runCmd finishes.
-	// If cmdFinishedCh were unbuffered, sending the final error would block
-	// (since no goroutine is receiving yet), which prevents runCmd from finishing.
-	// This means cmdWriter.Close() never runs, so the reader never gets EOF and hangs.
-	// With a buffered channel, the send doesn't block, allowing runCmd to finish,
-	// the writer to close, and the reader to eventually see EOF and exit.
-	//
-	// This is safe because we only send once, after the command finishes.
-	// Alternatively, we could manually close cmdWriter right after the command.
-	cmdFinishedCh := make(chan int, 1)
+	if err := cmd.Start(); err != nil {
+		return nil, fmt.Errorf("failed to start command: %w", err)
+	}
+
+	outputCh := make(chan outputLine, 10) // buffered to avoid blocking
```

Comment:
If the consumer gets very slow the command will block once these 10 entries get filled right? Just to make sure I understand.